File: src/passes/data_refs.cpp
// data-refs -- resolve constants that point at data.
//
// The sweep only decodes instructions, but the image it was given usually
// holds more than that. A constant operand landing inside the image, outside
// the range the sweep walked, is a pointer to data -- so say what is there
// instead of leaving a bare number.
#include "../pass.h"

#include "opcodes.hh"
#include "sleigh.hh"

#include <ostream>
#include <sstream>

namespace ddd {
namespace {

std::string escape(const std::string &text) {
  std::string out;
  for (char c : text) {
    switch (c) {
    case '\n':
      out += "\\n";
      break;
    case '\t':
      out += "\\t";
      break;
    case '"':
      out += "\\\"";
      break;
    case '\\':
      out += "\\\\";
      break;
    default:
      out.push_back(c);
    }
  }
  return out;
}

class DataRefs final : public Pass {
public:
  std::string name() const override { return "data-refs"; }
  std::string description() const override {
    return "resolve constants pointing into the image's data";
  }

  void run(SsaFunction &fn, PassContext &ctx) override {
    if (ctx.image == nullptr || ctx.image->empty()) {
      if (ctx.verbose)
        ctx.stream() << "  no image, skipping\n";
      return;
    }

    int resolved = 0;
    fn.for_each_op([&](SsaOp &op) {
      // Branch and call destinations are code; they are already shown as
      // block edges and would only add noise here.
      if (op.opc == ghidra::CPUI_BRANCH || op.opc == ghidra::CPUI_CBRANCH ||
          op.opc == ghidra::CPUI_CALL)
        return;

      for (size_t i = 0; i < op.ins.size(); ++i) {
        const SsaOperand &in = op.ins[i];
        if (!in.is_constant())
          continue;
        // Some constant operands are not values at all: the address space of a
        // LOAD/STORE, the userop index of a CALLOTHER.
        if (is_space_operand(op, i))
          continue;
        if (op.opc == ghidra::CPUI_CALLOTHER && i == 0)
          continue;

        // A load names its own width, and the fact that the program read
        // the address as data is better evidence than any heuristic: an ARM
        // literal pool sits *inside* .text, where is_data() says no.
        const bool loaded = op.opc == ghidra::CPUI_LOAD && i == 1;
        unsigned width = pointer_width(ctx);
        if (loaded) {
          if (op.out != nullptr) width = op.out->storage.size;
          else if (op.has_raw_output) width = op.raw_output.size;
        }

        std::string described =
            describe(*ctx.image, in.constant(), width, loaded);
        if (described.empty())
          continue;

        ctx.annotations->comment(op, described);
        ++resolved;
      }
    });

    if (ctx.verbose)
      ctx.stream() << "  " << resolved << " data reference(s)\n";
  }

private:
  // A constant is only worth resolving if it points somewhere that is data.
  //
  // Reading a word always "succeeds" anywhere inside the image, so that on its
  // own is no evidence at all -- with a base of 0 every small immediate would
  // come back as a pointer. Two things separate a real reference:
  //
  //   * a NUL-terminated printable run is self-validating, and is reported
  //     wherever it is found
  //   * anything else has to point past the end of everything disassembled.
  //     Small integers alias with the low addresses; the trailing data area
  //     does not.
  static unsigned pointer_width(const PassContext &ctx) {
    if (ctx.target == nullptr || ctx.target->translator == nullptr)
      return 8;
    return ctx.target->translator->getDefaultCodeSpace()->getAddrSize();
  }

  static std::string describe(const Image &image, uint64_t address,
                              unsigned width, bool loaded) {
    if (loaded ? !image.contains(address) : !image.is_data(address))
      return {};
    if (width == 0 || width > 8)
      return {};

    std::ostringstream os;
    os << "0x" << std::hex << address << " -> ";

    if (std::optional<std::string> text = image.read_string(address)) {
      os << '"' << escape(*text) << '"';
      return os.str();
    }

    if (!loaded && address < image.code_end())
      return {};

    std::optional<uint64_t> word = image.read_int(address, width);
    if (!word) {
      os << "data";
      return os.str();
    }

    os << "0x" << std::hex << *word;

    // One more hop, and no further: a literal pool entry is a pointer, and
    // the thing worth reading is what it points at, not the pointer.
    if (std::optional<std::string> text = image.read_string(*word))
      os << " -> \"" << escape(*text) << '"';
    else if (image.is_code(*word) && *word != 0)
      os << " (code)";

    return os.str();
  }
};

DDD_REGISTER_PASS(DataRefs);

} // namespace
} // namespace ddd

File: lua/plugins/analyses/data_refs.lua
-- data-refs -- resolve constants that point at data.
--
-- The sweep only decodes instructions, but the image it was given usually holds
-- more than that. A constant operand landing inside the image, outside the
-- range the sweep walked, is a pointer to data -- so say what is there instead
-- of leaving a bare number.
local ddd = require "ddd"

local escapes = { ["\n"] = "\\n", ["\t"] = "\\t", ['"'] = '\\"', ["\\"] = "\\\\" }

local function escape(text)
  return (text:gsub('[\n\t"\\]', escapes))
end

local function hex(value) return ("0x%x"):format(value) end

-- A constant is only worth resolving if it points somewhere that is data.
--
-- Reading a word always "succeeds" anywhere inside the image, so that on its
-- own is no evidence at all -- with a base of 0 every small immediate would
-- come back as a pointer. Two things separate a real reference:
--
--   * a NUL-terminated printable run is self-validating, and is reported
--     wherever it is found
--   * anything else has to point past the end of everything disassembled.
--     Small integers alias with the low addresses; the trailing data area does
--     not.
local function describe(ctx, address, width, loaded)
  if loaded then
    if not ctx:contains(address) then return nil end
  elseif not ctx:is_data(address) then
    return nil
  end

  if width == 0 or width > 8 then return nil end

  local text = ctx:read_string(address)
  if text then
    return ('%s -> "%s"'):format(hex(address), escape(text))
  end

  if not loaded and address < ctx.code_end then return nil end

  local word = ctx:read_int(address, width)
  if not word then return ("%s -> data"):format(hex(address)) end

  local described = ("%s -> %s"):format(hex(address), hex(word))

  -- One more hop, and no further: a literal pool entry is a pointer, and the
  -- thing worth reading is what it points at, not the pointer.
  local pointed = ctx:read_string(word)
  if pointed then
    return described .. (' -> "%s"'):format(escape(pointed))
  elseif word ~= 0 and ctx:is_code(word) then
    return described .. " (code)"
  end
  return described
end

-- Branch and call destinations are code; they are already shown as block edges
-- and would only add noise here.
local skip = { BRANCH = true, CBRANCH = true, CALL = true }

ddd.workflow "readability" {
  passes = function(scope)
    scope.pass "data-refs" {
      description = "resolve constants pointing into the image's data",

      before = function(_, ctx) ctx.resolved = 0 end,

      each_op = function(op, _, ctx)
        if skip[op.opcode] then return end

        for index, operand in op:inputs() do
          -- Some constant operands are not values at all: the address space of
          -- a LOAD or STORE, the userop index of a CALLOTHER.
          local structural = operand.is_space
            or (op.opcode == "CALLOTHER" and index == 1)

          if operand.is_constant and not structural then
            -- A load names its own width, and the fact that the program read
            -- the address as data is better evidence than any heuristic: an
            -- ARM literal pool sits *inside* .text, where is_data() says no.
            local loaded = op.opcode == "LOAD" and index == 2
            local width = ctx.pointer_size
            if loaded then
              if op.out then
                width = op.out.size
              elseif op.raw_out then
                width = op.raw_out.size
              end
            end

            local described = describe(ctx, operand.constant, width, loaded)
            if described then
              ctx:comment(op, described)
              ctx.resolved = ctx.resolved + 1
            end
          end
        end
      end,

      after = function(_, ctx)
        if ctx.verbose then
          ctx:log(("  %d data reference(s)"):format(ctx.resolved))
        end
      end,
    }
  end,
}
